File: sdk/python/src/openlit/guard/prompt_injection.py
"""
Prompt injection & jailbreak detection guard.

Fast regex patterns catch known injection signatures.  An optional
user-provided ``Callable[[str], float]`` handles ambiguous cases.
Phases: preflight only.
"""

from __future__ import annotations

import re
from typing import Callable, List, Optional, Tuple

from openlit.guard._base import Guard, GuardPhase, GuardResult

# (label, pattern, weight)  -- weight contributes to the aggregated score.
_INJECTION_PATTERNS: List[Tuple[str, re.Pattern, float]] = []


def _p(label: str, pattern: str, weight: float = 0.5) -> None:
    _INJECTION_PATTERNS.append((label, re.compile(pattern, re.IGNORECASE), weight))


# ---- Instruction override ----
_p(
    "instruction-override",
    r"ignore\s+(?:all\s+)?(?:previous|above|prior)\s+(?:instructions|prompts|rules)",
    0.9,
)
_p(
    "instruction-override-2",
    r"disregard\s+(?:all\s+)?(?:previous|above|prior)\s+(?:instructions|context)",
    0.9,
)
_p("new-instructions", r"(?:new|updated|revised)\s+instructions\s*:", 0.7)
_p("do-anything-now", r"(?:DAN|do\s+anything\s+now)\s+mode", 0.95)
_p("jailbreak-keyword", r"jailbreak(?:ed|ing)?", 0.8)

# ---- System prompt extraction ----
_p(
    "system-prompt-leak",
    (
        r"(?:show|reveal|display|print|output|repeat|tell\s+me)"
        r"\s+(?:your|the|me\s+your)"
        r"\s*(?:system|initial|original|hidden)"
        r"\s+(?:prompt|instructions|message)"
    ),
    0.85,
)
_p(
    "system-prompt-leak-2",
    r"what\s+(?:are|were)\s+your\s+(?:system|initial|original)\s+(?:instructions|prompt)",
    0.8,
)
_p(
    "system-prompt-leak-3",
    r"(?:show|reveal|display|print|output|repeat)\s+(?:me\s+)?your\s+(?:system\s+)?prompt",
    0.8,
)

# ---- Role impersonation ----
_p(
    "role-play",
    (
        r"(?:you\s+are\s+now|act\s+as|pretend\s+(?:to\s+be|you\s+are)|roleplay\s+as)"
        r"\s+(?:a\s+)?(?:hacker|evil|malicious|unrestricted|unfiltered)"
    ),
    0.85,
)
_p("developer-mode", r"(?:developer|debug|admin|god|sudo|root)\s+mode", 0.7)

# ---- Encoding bypass ----
_p("base64-injection", r"(?:decode|base64|eval)\s*\(", 0.6)
_p("markdown-injection", r"\[.*?\]\((?:javascript|data):", 0.8)

# ---- Delimiter abuse ----
_p(
    "delimiter-abuse",
    r"={5,}|<\|(?:im_start|system|endoftext)\|>|###\s*(?:system|instruction)",
    0.7,
)


class PromptInjection(Guard):
    """
    Detects prompt injection and jailbreak attempts.

    Parameters
    ----------
    action : str
        ``"deny"`` (block the call) or ``"warn"`` (emit event only).
    threshold : float
        Minimum aggregated score to trigger the configured action (0.0--1.0).
    classifier : callable, optional
        ``Callable[[str], float]`` returning an injection probability (0--1).
        Used as a fallback when regex patterns don't match.
    """

    name = "prompt_injection"
    phases = (GuardPhase.PREFLIGHT,)

    def __init__(
        self,
        action: str = "deny",
        threshold: float = 0.5,
        classifier: Optional[Callable[[str], float]] = None,
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)
        self._threshold = threshold
        self._classifier = classifier

    def evaluate(self, text: str) -> GuardResult:
        matched_labels: List[str] = []
        max_weight = 0.0

        for label, pattern, weight in _INJECTION_PATTERNS:
            if pattern.search(text):
                matched_labels.append(label)
                max_weight = max(max_weight, weight)

        score = max_weight

        if not matched_labels and self._classifier is not None:
            score = self._classifier(text)

        if score >= self._threshold:
            classification = (
                ", ".join(matched_labels) if matched_labels else "classifier"
            )
            return GuardResult(
                action=self._action,
                score=round(score, 3),
                guard_name=self.name,
                classification=classification,
                explanation=f"Prompt injection detected (score={score:.2f}): {classification}",
            )

        return GuardResult(guard_name=self.name, score=round(score, 3))

File: sdk/typescript/src/guard/prompt-injection.ts
/**
 * Prompt injection & jailbreak detection guard.
 *
 * Fast regex patterns catch known injection signatures. An optional
 * user-provided classifier handles ambiguous cases.
 *
 * Patterns must stay in sync with: sdk/python/src/openlit/guard/prompt_injection.py
 */

import { Guard, GuardPhase, GuardResult, makeGuardResult, GuardOptions } from './base';

interface InjectionPattern {
  label: string;
  regex: RegExp;
  weight: number;
}

const INJECTION_PATTERNS: InjectionPattern[] = [
  // ---- Instruction override ----
  { label: 'instruction-override', regex: /ignore\s+(?:all\s+)?(?:previous|above|prior)\s+(?:instructions|prompts|rules)/i, weight: 0.9 },
  { label: 'instruction-override-2', regex: /disregard\s+(?:all\s+)?(?:previous|above|prior)\s+(?:instructions|context)/i, weight: 0.9 },
  { label: 'new-instructions', regex: /(?:new|updated|revised)\s+instructions\s*:/i, weight: 0.7 },
  { label: 'do-anything-now', regex: /(?:DAN|do\s+anything\s+now)\s+mode/i, weight: 0.95 },
  { label: 'jailbreak-keyword', regex: /jailbreak(?:ed|ing)?/i, weight: 0.8 },
  // ---- System prompt extraction ----
  { label: 'system-prompt-leak', regex: /(?:show|reveal|display|print|output|repeat|tell\s+me)\s+(?:your|the|me\s+your)\s*(?:system|initial|original|hidden)\s+(?:prompt|instructions|message)/i, weight: 0.85 },
  { label: 'system-prompt-leak-2', regex: /what\s+(?:are|were)\s+your\s+(?:system|initial|original)\s+(?:instructions|prompt)/i, weight: 0.8 },
  { label: 'system-prompt-leak-3', regex: /(?:show|reveal|display|print|output|repeat)\s+(?:me\s+)?your\s+(?:system\s+)?prompt/i, weight: 0.8 },
  // ---- Role impersonation ----
  { label: 'role-play', regex: /(?:you\s+are\s+now|act\s+as|pretend\s+(?:to\s+be|you\s+are)|roleplay\s+as)\s+(?:a\s+)?(?:hacker|evil|malicious|unrestricted|unfiltered)/i, weight: 0.85 },
  { label: 'developer-mode', regex: /(?:developer|debug|admin|god|sudo|root)\s+mode/i, weight: 0.7 },
  // ---- Encoding bypass ----
  { label: 'base64-injection', regex: /(?:decode|base64|eval)\s*\(/i, weight: 0.6 },
  { label: 'markdown-injection', regex: /\[.*?\]\((?:javascript|data):/i, weight: 0.8 },
  // ---- Delimiter abuse ----
  { label: 'delimiter-abuse', regex: /={5,}|<\|(?:im_start|system|endoftext)\|>|###\s*(?:system|instruction)/i, weight: 0.7 },
];

export interface PromptInjectionOptions extends GuardOptions {
  threshold?: number;
  classifier?: (text: string) => number;
}

export class PromptInjection extends Guard {
  readonly name = 'prompt_injection';
  readonly phases = [GuardPhase.PREFLIGHT];

  private readonly _threshold: number;
  private readonly _classifier?: (text: string) => number;

  constructor(opts: PromptInjectionOptions = {}) {
    super({ action: opts.action ?? 'deny', maxScanLength: opts.maxScanLength });
    this._threshold = opts.threshold ?? 0.5;
    this._classifier = opts.classifier;
  }

  evaluate(text: string): GuardResult {
    const matchedLabels: string[] = [];
    let maxWeight = 0;

    for (const { label, regex, weight } of INJECTION_PATTERNS) {
      if (regex.test(text)) {
        matchedLabels.push(label);
        if (weight > maxWeight) {
          maxWeight = weight;
        }
      }
    }

    let score = maxWeight;

    if (matchedLabels.length === 0 && this._classifier) {
      score = this._classifier(text);
    }

    if (score >= this._threshold) {
      const classification = matchedLabels.length > 0 ? matchedLabels.join(', ') : 'classifier';
      return makeGuardResult({
        action: this._action,
        score: Math.round(score * 1000) / 1000,
        guardName: this.name,
        classification,
        explanation: `Prompt injection detected (score=${score.toFixed(2)}): ${classification}`,
      });
    }

    return makeGuardResult({ guardName: this.name, score: Math.round(score * 1000) / 1000 });
  }
}
